### PR TITLE
Added `VideoFrame::GetPixelType()`.

### LINKED ISSUE #317: `VideoFrame` should provide its color format
A `VideoFrame` should bring with it the necessary data to handle it. You can already query its width (advancement- and validity-wise) as well as height. You should also be able to query the information what exactly the bytes describe, i.e., the color format.

Often, you also have access to the `IClip` to get the color format from. But when retrieving a `VideoFrame` with `env->propGetFrame()`, there is no defined relationship to an `IClip`.

Not being able to get the associated color format when getting a frame out of a frame prop is a problem for my Rust crate.

### PR DESCRIPTION
**Pull request by Enyium**
Closes #317.

Don't forget to change the line in `struct AVS_Linkage` with the `TEMPORARY ENTRY` comment when resolving the conflict there that would happen anyway after merging #331.

### CODE REVIEW

**Comment by pinterf** [comment]
Yep, SubFrame thing seems also good, this is something I imagined. I'll check CombinePlanes I mentioned but you have not found, maybe I did not remember correctly.

**Comment by Enyium** [comment]
BTW, does the following code from the codebase have frame writability problems, by any chance?

```cpp
PVideoFrame FilterInfo::GetFrame(int n, IScriptEnvironment* env)
{
  PVideoFrame frame = vii.HasVideo() ? child->GetFrame(n, env) : env->NewVideoFrame(vi);

  if ( !vii.HasVideo() ) {
    memset(frame->GetWritePtr(), 0, frame->GetPitch()*frame->GetHeight()); // Blank frame
  }

#if defined(AVS_WINDOWS) && !defined(NO_WIN_GDI)
  HDC hdcAntialias = antialiaser.GetDC();
  if (!hdcAntialias)
    return frame;
#else
  if (current_font == nullptr)
    return frame;
#endif
//..................................
  return frame;
}
```

**Comment by pinterf** [comment]
I don't think so. You mean, the memset section? That branch runs right after a NewVideoFrame. Later, before the actual write there is MakeWritable.

**Comment by pinterf** [comment]
I've found the CombinePlanes issue. Test script:

```
# Copy planes between planar RGB(A) and YUV(A) without any conversion
# yuv 4:4:4 <-> planar rgb
BlankClip(pixel_type="YV24")
source = last.ConvertBits(32) # 4:4:4
cast_to_planarrgb = CombinePlanes(source, planes="RGB", source_planes="YUV", pixel_type="RGBPS")
# get back a clip identical with "source"
#cast_to_yuv = CombinePlanes(cast_to_planarrgb, planes="YUV", source_planes="RGB", pixel_type="YUV444PS")
cast_to_planarrgb
```

This is the place when the planes are simply swapped:
https://github.com/AviSynth/AviSynthPlus/blob/master/avs_core/filters/planeswap.cpp#L832

`SubFramePlanarA` and `SubFramePlanar` is used.
https://github.com/AviSynth/AviSynthPlus/blob/master/avs_core/filters/planeswap.cpp#L881

But if source is RGB(A) and target is YUV(A) (or vice versa) then we have to use amend for VideoFrame pixel_type, since SubFramePlanarA and SubFramePlanar won't detect the change.
Basically this is the condition: `(vi_src.IsRGB() && vi.IsYUV()) || (vi_src.IsYUV() && vi.IsRGB())`
SubFrame is O.K. here.

Or simply we do not enter into that branch if they are of different basic color spaces.

**Comment by Enyium** [comment]
I changed that one spot and rebased onto the latest revision from master, removing the `TEMPORARY ENTRY` comment.

**Comment by Enyium** [comment]
I put the new functions in `avisynth.h` *after* the destructor `~VideoFrame()` to not disturb vtable order. I guess this was also done incorrectly in the past when adding `IsPropertyWritable()` before `~VideoFrame()`. Maybe it's not that relevant because the destructor shouldn't be called by plugin code. But it has `AVS_Linkage` code: `AVS_BakedCode( AVS_LinkCall_Void(VideoFrame_DESTRUCTOR)() )`.

Maybe MSVC already reordered the destructor for a long time (as it does for method overloads), so it may be that it doesn't matter for *that* reason.

**Comment by pinterf** [comment]
I do not get the logic in how vtable order can be governed, I couldn't find a white paper on how vtable order is standardized, at least within MSVC.

This was the case with NewVideoFrame. A different name, NewVideoFrameP had to be introduced instead. I spent days on the topic at that time but finally had to make the name change. I had to rename GetVar versions (to GetVarBool, GetVarInt, etc...) for probably the same reason, when I put them to the public interface from IScriptEnvironment2.

Then I put a warning in Avisynth.h:

_Note: do not declare existing names like 'NewVideoFrame' again with different parameters since MSVC will reorder it
in the vtable and group it together with the first NewVideoFrame variant.
This results in shifting all vtable method pointers after NewVideoFrame and breaks all plugins who expect the old order.
E.g. ApplyMessage will be called instead of GetAVSLinkage_

**Comment by Enyium** [comment]
> I couldn't find a white paper on how vtable order is standardized, at least within MSVC.

It isn't standardized. There's obviously just a de-facto standard. And the AviSynth+ ecosystem depends on it. https://stackoverflow.com/questions/30206215/how-are-c-vtable-methods-ordered-in-practice

> Then I put a warning in Avisynth.h:

Yes, I saw that comment.

**Comment by pinterf** [comment]
(MSVC, Windows) As long as it is compatible with COM interop, the order of declaration is sequential. 
https://stackoverflow.com/questions/61542063/inversion-of-generated-vtable-functions-order-for-functions-with-the-same-name
But if not, overloads in the same class are grouped together in the vtable in the reverse order of declaration.

**Comment by pinterf** [comment]
> I put the new functions in `avisynth.h` _after_ the destructor `~VideoFrame()` to not disturb vtable order. I guess this was also done incorrectly in the past when adding `IsPropertyWritable()` before `~VideoFrame()`. Maybe it's not that relevant because the destructor shouldn't be called by plugin code. But it has `AVS_Linkage` code: `AVS_BakedCode( AVS_LinkCall_Void(VideoFrame_DESTRUCTOR)() )`.

(I'm a little bit feeling lost) Isn't it only IScriptEnvironment's virtual table order is important from plugins' point of view, and the functions put into the AVS_Linkage and called through AVS_LinkCall are safe?

**Comment by Enyium** [comment]
> (MSVC, Windows) As long as it is compatible with COM interop, the order of declaration is sequential.
> https://stackoverflow.com/questions/61542063/inversion-of-generated-vtable-functions-order-for-functions-with-the-same-name
> But if not, overloads in the same class are grouped together in the vtable in the reverse order of declaration.

Good link.

> (I'm a little bit feeling lost) Isn't it only IScriptEnvironment's virtual table order is important from plugins' point of view, and the functions put into the AVS_Linkage and called through AVS_LinkCall are safe?

I just tested it. The only classes where wild reordering of functions (including in the `#ifdef BUILDING_AVSCORE` sections) breaks AviSynth+ are `IClip` and `IScriptEnvironment`. Reordering in all other classes and `struct VideoInfo` didn't cause problems.

Reordering fields of classes and structs nearly always causes problems, though (except in `union`s). Strangely, it isn't a problem in `GenericVideoFilter`.

**Comment by pinterf** [comment]
I think this is safe now. 

I think we'd need a C interface counterpart of 
```
  int           (VideoFrame::*VideoFrame_GetPixelType)() const;
  void          (VideoFrame::*VideoFrame_AmendPixelType)(int new_pixel_type);
```
We do not allow (though it seems possible) to access AVS_VideoFrame struct members directly.

**Comment by Enyium** [comment]
Sorry, I don't seem to get along with the C interface.

**Comment by pinterf** [comment]
O.K., no problem, I just wanted to write that I should review the C interface more, not just the current extension.
